Add transpose_coords option to DataArray.transpose

phausamann opened the pull request:
- [x] Closes #1856 
 - [x] Tests added (for all bug fixes or enhancements)
 - [x] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

I've added the option `transpose_coords` to `DataArray.transpose` as described in #1856. It's `False` by default, as it breaks a couple of tests otherwise (`TestConcatDataset.test_concat`, `test_apply_output_core_dimension` and `test_dot`). I'm not sure whether fixing these tests to work with the new behavior should be part of the PR.

pep8speaks commented:
Hello @phausamann! Thanks for updating the PR.

 - There are no PEP8 issues in the file [`xarray/core/dataarray.py`](https://github.com/phausamann/xarray/blob/5b7105bef48883204a0f68fba6f1a98debcc0062/xarray/core/dataarray.py) !

 - There are no PEP8 issues in the file [`xarray/tests/test_dataarray.py`](https://github.com/phausamann/xarray/blob/5b7105bef48883204a0f68fba6f1a98debcc0062/xarray/tests/test_dataarray.py) !

 - In the file [`xarray/tests/test_dataset.py`](https://github.com/phausamann/xarray/blob/5b7105bef48883204a0f68fba6f1a98debcc0062/xarray/tests/test_dataset.py), following                are the PEP8 issues :
> [Line 3803:23](https://github.com/phausamann/xarray/blob/5b7105bef48883204a0f68fba6f1a98debcc0062/xarray/tests/test_dataset.py#L3803): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent

shoyer commented:
As a default, I think we want `transpose_coords=None`.

If `transpose_coords is None` and there are coordinates to transpose, we can issue a `FutureWarning` about how the default behavior of `transpose_coords` will switch to `True` in the future.

> I'm not sure whether fixing these tests to work with the new behavior should be part of the PR.

Yes, this would be ideal. It would be fine to use `transpose_coords=False` explicitly in tests, but we should silence all warnings.

phausamann commented:
I've updated the method to issue the warning and fixed the failing tests (which was actually a bug in my  implementation).

> It would be fine to use `transpose_coords=False` explicitly in tests, but we should silence all warnings.

I'm not exactly sure how this works, there are a lot of tests where the warning is issued, especially calls to `DataArray.T`. Should I update all of these tests to `transpose(..., transpose_coords=False)`?

phausamann commented on xarray/core/dataarray.py:
```diff
@@ -1381,8 +1383,29 @@ def transpose(self, *dims):
+                                 'permuted array dimensions (%s)'
+                                 % (dims, tuple(self.dims)))
+
+        transpose_coords = kwargs.pop('transpose_coords', None)
```
Should we raise an error when keyword args other than `transpose_coords` are being passed? `kwargs` is necessary for Python 2 compatibility.

shoyer replied:
if we wait 40 days before mergin this, we wouldn't have to worry about Python 2 compatibility at all :)

But yes, for now:
```
if kwargs:
    raise TypeError('unrecognized keyword arguments')
```

shoyer commented:
> I'm not exactly sure how this works, there are a lot of tests where the warning is issued, especially calls to `DataArray.T`. Should I update all of these tests to `transpose(..., transpose_coords=False)`?

Yes, that would probably be the right call. Or maybe even switch to `.transpose(transpose_coords=True)` to future proof the behavior?

We try to avoid issuing warnings in xarray's own test suite so we notice issues (e.g., deprecations in upstream packages) when they arise. This also gives us some estimate of how much turmoil a change will cause for users -- if we end up with lots of internal warnings, then users will probably notice, too.

phausamann commented on xarray/core/groupby.py:
```diff
@@ -473,7 +473,7 @@ def lookup_order(dimension):
 
         new_order = sorted(stacked.dims, key=lookup_order)
-        return stacked.transpose(*new_order)
+        return stacked.transpose(*new_order, transpose_coords=True)
```
Should this be `False`?

shoyer replied:
We should probably do the same thing for this transpose in groupby that we did for transpose() in general -- we should have a keyword argument to control this behavior, e.g., perhaps `ds.groupby('something', restore_coord_dims=True).mean()`?

We definitely don't want to issue warnings internally that users can't control. `groupby()` using `transpose()` internally is an implementation detail that isn't part of xarray's public API.

phausamann commented:
I've change all the `tranpose` calls in the tests to `transpose_coords=True`. Other than that, there's some `tranpose` call in `plot` and `groupby` that I've updated, but I'm not sure whether they should be `True`, `False` or not updated at all (see above).

phausamann commented:
I've added a `restore_coord_dims` parameter to `groupby`, `groupby_bins` and `resample` and eliminated all warnings from the tests. The build failure seems to be inherited from the master branch.

phausamann commented:
@shoyer is this ready to be merged?

StanczakDominik commented:
Hi, just popping in - perhaps rebasing on top of the master branch would help move this along?

shoyer commented:
I just merged in master, let's see how that goes

phausamann commented:
Looks like the build failures were only temporary, all the checks are passing now. I think this is good to go.

shoyer commented:
OK, in it goes!

Thanks for your patience @phausamann !
